File: iutil/xd.js

function XD(memory, bytes) {
  function ascii_to_hex(n, pad_to) {
    s = n.toString(16);
    while (s.length < pad_to) {
      s = '0' + s;
    }
    return s;
  }

  var i = 0;
  var j = i;
  var m = memory;
  var last_ascii = 0;
  var ascii = '';
  var a_byte = '';
  var xd = '';
  for (i = 0; i < bytes; ++i) {
    if ((i % 16) == 0) {
      xd += ascii_to_hex(i, 8);
    }
    xd += ascii_to_hex(m.charCodeAt(i), 2);
    if ((i % 2) == 1) {
      xd += ' ';
    }
    if ((i % 16) == 15) {
      xd += ' ';
      for (j = last_ascii; j <= i; ++j) {
        a_byte = m.charCodeAt(j);
        if (' ' < a_byte && a_byte <= '~') {
          ascii = a_byte;
        } else {
          ascii = '.';
        }
        xd += ascii;
      }
      last_ascii = i
      xd += '\n';
    }
  }
}

File: iutil/xd.c

static void xd_byte(unsigned char b) {
    static char hexes[] = "0123456789abcedf";
    printf("%c%c", hexes[b >> 4], hexes[b & 0xf]);
}

void XD(void *memory, unsigned long long bytes) {
    unsigned long long i;
    unsigned long long j;
    unsigned char *m = memory;
    unsigned long long last_ascii = 0;
    unsigned char ascii;
    for (i = 0; i < bytes; ++i) {
        if ((i % 16) == 0) {
            printf("%08llx: ", i);
        }
        printf("%02hhx", m[i]);
        if ((i % 2) == 1) {
            printf(" ");
        }
        if ((i % 16) == 15) {
            printf(" ");
            for (j = last_ascii; j <= i; ++j) {
                if (' ' < m[j] && m[j] <= '~') {
                    ascii = m[j];
                } else {
                    ascii = '.';
                }
                printf("%c", (char)ascii);
            }
            last_ascii = i;
            printf("\n");
        }
    }
    if ((i % 16) != 15) {
        printf("\n");
    }
}
